test/distributed/fsdp/test_fsdp_fine_tune.py

```diff
@@ -151,6 +151,68 @@ def _assert_post_backward_reshard_count(step_idx, num_steps):
                     _assert_post_backward_reshard_count(step_idx, num_steps)
                     post_backward_reshard_count = 0
 
+    def _init_multi_traversal_module(self) -> nn.Module:
+        torch.manual_seed(42)
+
+        class TestModule(nn.Module):
+            def __init__(self):
+                super().__init__()
+                self.layer_0 = nn.Linear(5, 5, device="cuda")
+                self.layer_no_grad = nn.Linear(5, 5, device="cuda")
+                self.layer_with_grad = nn.Linear(5, 5, device="cuda")
+                self.layer_no_grad.requires_grad_(False)
+
+            def forward(self, x):
+                # Layer `layer_no_grad` and `layer_with_grad` are called
+                # multiple times, IOW, their parameters are used multiple times
+                # during forward pass.
+                x = self.layer_0(x)
+                for _ in range(10):
+                    x = self.layer_no_grad(self.layer_with_grad(x))
+                return x
+
+        return TestModule()
+
+    @skip_if_lt_x_gpu(2)
+    def test_backward_hooks_multi_traversal(self):
+        """
+        Tests that the backward hooks do reshard / unshard correctly in the case
+        of same parameters being used mutliple times during forward pass.
+        """
+        self.run_subtests(
+            {
+                "sharding_strategy": [
+                    ShardingStrategy.FULL_SHARD,
+                    ShardingStrategy.SHARD_GRAD_OP,
+                    ShardingStrategy.NO_SHARD,
+                ],
+                "use_orig_params": [False, True],
+                "inp_requires_grad": [False, True],
+            },
+            self._test_backward_hooks_multi_traversal,
+        )
+
+    def _test_backward_hooks_multi_traversal(
+        self,
+        sharding_strategy: ShardingStrategy,
+        use_orig_params: bool,
+        inp_requires_grad: bool,
+    ):
+        seq = self._init_multi_traversal_module()
+        policy = ModuleWrapPolicy({nn.Linear})
+        seq = FSDP(
+            seq,
+            auto_wrap_policy=policy,
+            sharding_strategy=sharding_strategy,
+            use_orig_params=use_orig_params,
+        )
+        inp = torch.randn((8, 5), device="cuda", requires_grad=inp_requires_grad)
+        output = seq(inp)
+        output.sum().backward()
+
+        # No assertion here. The case exiting normally confirms that we do shard
+        # / unshard correctly.
```

Comment:
Thanks for the great unit test!

I think it would be better if we checked for numeric correctness too. I am not sure if DDP supports this case, but if so, then one option is to compare against DDP for numerics.

If DDP does not support this, then we can also compare against a local model, where the local model processes a global batch with batch size equal to `self.world_size * local_batch_size`, while the FSDP model processes a batch with size `local_batch_size`. Then, we can manually divide the local model's gradients by `self.world_size`, and we should expect numeric parity, which can be checked by running a training loop and comparing the forward outputs (or loss).

Reply:
Thanks for the feedback.

I reworked the test by referring to the implementation of `test_parity_with_ddp`. Now the test compares the loss to DDP's loss for correctness.